### PR TITLE
storage: Readonly crypto improvements

### PR DESCRIPTION
**Pull request by mvollmer**

### CODE REVIEW

File: pkg/storaged/crypto/keyslots.jsx [hunk 1]
```diff
@@ -72,22 +75,51 @@ export function clevis_recover_passphrase(block, just_type) {
             .then(output => output.trim());
 }
 
-function clevis_unlock(block) {
+async function clevis_unlock(client, block, luksname, readonly) {
     const dev = decode_filename(block.Device);
-    const clear_dev = "luks-" + block.IdUUID;
-    return cockpit.spawn(["clevis", "luks", "unlock", "-d", dev, "-n", clear_dev],
-                         { superuser: true });
+    const clear_dev = luksname || "luks-" + block.IdUUID;
```
Comment (cockpituous) [inline pkg/storaged/crypto/keyslots.jsx:80]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-20148-20240307-101024-121dd943-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/block/format-dialog.jsx [hunk 2]
```diff
@@ -494,6 +494,8 @@
                 if (is_encrypted(vals)) {
                     let opts = [];
                     if (is_filesystem(vals)) {
+                        if (vals.mount_options?.ro)
+                            opts.push("readonly");
```
Comment (cockpituous) [inline pkg/storaged/block/format-dialog.jsx:498]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-20148-20240307-140056-b1476764-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/block/format-dialog.jsx [hunk 3]
```diff
@@ -572,17 +574,25 @@
...
-                function maybe_unlock() {
+                async function maybe_unlock() {
                     const content_block = client.blocks_cleartext[path];
-                    if (content_block)
+                    if (content_block) {
+                        if (content_block.ReadOnly) {
+                            console.log("REOPEN read/write");
+                            const block_crypto = client.blocks_crypto[path];
+                            await block_crypto.Lock({});
+                            await unlock_with_type(client, block, vals.old_passphrase, existing_passphrase_type, false);
```
Comment (cockpituous) [inline pkg/storaged/block/format-dialog.jsx:584]: These 5 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-20148-20240307-140056-b1476764-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/block/format-dialog.jsx [hunk 4]
```diff
@@ -572,17 +574,25 @@
...
-                                dlg.set_values({ needs_explicit_passphrase: true });
-                                return Promise.reject(error);
-                            })
-                            .then(() => client.blocks_cleartext[path]));
+                    try {
+                        await unlock_with_type(client, block, vals.old_passphrase, existing_passphrase_type, false);
+                        return client.blocks_cleartext[path];
+                    } catch (error) {
+                        dlg.set_values({ needs_explicit_passphrase: true });
+                        throw error;
```
Comment (cockpituous) [inline pkg/storaged/block/format-dialog.jsx:594]: These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-20148-20240307-140056-b1476764-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/block/format-dialog.jsx [hunk 5]
```diff
@@ -644,6 +654,16 @@
                     if (is_encrypted(vals))
                         remember_passphrase(new_block, vals.passphrase);
 
+                    if (is_encrypted(vals) && is_filesystem(vals) && vals.mount_options?.ro) {
+                        console.log("REOPEN readonly")
+                        const block_crypto = await client.wait_for(() => block_crypto_for_block(path));
+                        await block_crypto.Lock({});
+                        if (vals.passphrase)
+                            await block_crypto.Unlock(vals.passphrase, { "read-only": { t: "b", v: true } });
```
Comment (cockpituous) [inline pkg/storaged/block/format-dialog.jsx:662]: These 6 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-20148-20240307-140056-b1476764-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/block/format-dialog.jsx [hunk 6]
```diff
@@ -644,6 +654,16 @@
...
+                        else
+                            await unlock_with_type(client, block, vals.old_passphrase, existing_passphrase_type, true);
```
Comment (cockpituous) [inline pkg/storaged/block/format-dialog.jsx:664]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-20148-20240307-140056-b1476764-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: pkg/storaged/crypto/keyslots.jsx (hunk 1 above)
Comment (cockpituous) [inline pkg/storaged/crypto/keyslots.jsx:80]: This added line is not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-20148-20240307-140056-b1476764-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Comment by mvollmer** [comment]
We don't currently get coverage reports as comments here. Results as of ee393f3788596c3a9b4022e687113361476c3076: https://cockpit-logs.us-east-1.linodeobjects.com/pull-20148-ee393f37-20240318-154529-fedora-39-devel/Coverage/lcov/github-pr.diff.gcov.html

**Comment by mvollmer** [comment]
test/verify/check-storage-luks TestStorageLuks.testFormatReadOnly runs out of memory on rhel-9-4.

File: pkg/storaged/block/format-dialog.jsx [hunk 7]
```diff
@@ -572,17 +576,24 @@ function format_dialog_internal(client, path, start, size, enable_dos_extended,
...
-                                dlg.set_values({ needs_explicit_passphrase: true });
-                                return Promise.reject(error);
-                            })
-                            .then(() => client.blocks_cleartext[path]));
+                    try {
+                        await unlock_with_type(client, block, vals.old_passphrase, existing_passphrase_type, false);
+                        return client.blocks_cleartext[path];
+                    } catch (error) {
+                        dlg.set_values({ needs_explicit_passphrase: true });
+                        throw error;
```
Comment (cockpituous) [inline pkg/storaged/block/format-dialog.jsx:595]: These 3 added lines are not executed by any test.

File: pkg/storaged/crypto/keyslots.jsx [hunk 8]
```diff
@@ -72,22 +76,51 @@ export function clevis_recover_passphrase(block, just_type) {
...
```
Comment (cockpituous) [inline pkg/storaged/crypto/keyslots.jsx:81]: This added line is not executed by any test.

File: pkg/storaged/filesystem/mounting-dialog.jsx [hunk 9]
```diff
@@ -253,18 +272,17 @@ export function mounting_dialog(client, block, mode, forced_options, subvol) {
...
-                          {
-                              visible: vals => vals.needs_explicit_passphrase,
-                              validate: val => !val.length && _("Passphrase cannot be empty"),
-                          })
-            ]);
+        fields = fields.concat([
+            PassInput("passphrase", _("Passphrase"),
+                      {
+                          visible: vals => vals.needs_explicit_passphrase,
+                          validate: val => !val.length && _("Passphrase cannot be empty"),
```
Comment (cockpituous) [inline pkg/storaged/filesystem/mounting-dialog.jsx:283]: This added line is not executed by any test.

File: pkg/storaged/filesystem/mounting-dialog.jsx [hunk 10]
```diff
@@ -341,10 +372,13 @@
                         opts = opts.concat(forced_options);
                     if (vals.mount_options?.extra)
                         opts = opts.concat(parse_options(vals.mount_options.extra));
+                    // XXX - take subvols into account.
+                    const crypto_unlock_readonly = vals.mount_options?.ro ?? opt_ro;
```
Comment (cockpituous) [inline pkg/storaged/filesystem/mounting-dialog.jsx:376]: This added line is not executed by any test.

File: pkg/storaged/block/format-dialog.jsx [hunk 11]
```diff
@@ -516,17 +520,24 @@ function format_dialog_internal(client, path, start, size, enable_dos_extended,
...
```
Comment (cockpituous) [inline pkg/storaged/block/format-dialog.jsx:539]: These 3 added lines are not executed by any test.

File: pkg/storaged/crypto/keyslots.jsx (hunk 8 above)
Comment (cockpituous) [inline pkg/storaged/crypto/keyslots.jsx:81]: This added line is not executed by any test.

File: pkg/storaged/filesystem/mounting-dialog.jsx [hunk 12]
```diff
@@ -314,18 +333,17 @@ export function mounting_dialog(client, block, mode, forced_options, subvol) {
...
-                          {
-                              visible: vals => vals.needs_explicit_passphrase,
-                              validate: val => !val.length && _("Passphrase cannot be empty"),
-                          })
-            ]);
+        fields = fields.concat([
+            PassInput("passphrase", _("Passphrase"),
+                      {
+                          visible: vals => vals.needs_explicit_passphrase,
+                          validate: val => !val.length && _("Passphrase cannot be empty"),
```
Comment (cockpituous) [inline pkg/storaged/filesystem/mounting-dialog.jsx:344]: This added line is not executed by any test.

File: pkg/storaged/filesystem/mounting-dialog.jsx [hunk 13]
```diff
@@ -399,10 +432,13 @@ export function mounting_dialog(client, block, mode, forced_options, subvol) {
                         opts = opts.concat(forced_options);
                     if (vals.mount_options?.extra)
                         opts = opts.concat(parse_options(vals.mount_options.extra));
+                    // XXX - take subvols into account.
+                    const crypto_unlock_readonly = vals.mount_options?.ro ?? opt_ro;
```
Comment (cockpituous) [inline pkg/storaged/filesystem/mounting-dialog.jsx:436]: This added line is not executed by any test.

File: pkg/storaged/filesystem/mounting-dialog.jsx [hunk 14]
```diff
@@ -243,16 +243,18 @@ export function mounting_dialog(client, block, mode, forced_options, subvol) {
...
+                try {
+                    await unlock_with_type(client, block, passphrase, passphrase_type);
+                    return await client.wait_for(() => client.blocks_cleartext[block.path]);
+                } catch (error) {
+                    dlg.set_values({ needs_explicit_passphrase: true });
+                    throw error;
+                }
             } else
-                return Promise.resolve();
+                return block;
```
[the remark this answers is not in the document]
Comment (mvollmer) [reply pkg/storaged/filesystem/mounting-dialog.jsx]: The exception is caught by the dialog machinery and turned into an error in the dialog itself.

(Exceptions thrown by async functions are equivalent to rejected Promises.)

File: pkg/storaged/block/format-dialog.jsx (hunk 11 above)
[the remark this answers is not in the document]
Comment (mvollmer) [reply pkg/storaged/block/format-dialog.jsx:539]: This is also caught by the dialog machinery.

File: pkg/storaged/crypto/keyslots.jsx [hunk 15]
```diff
@@ -72,22 +76,51 @@ export function clevis_recover_passphrase(block, just_type) {
...
+        await crypto.Unlock(passphrase, unlock_options);
+        return;
+    }
+
+    await cockpit.spawn(["clevis", "luks", "unlock", "-d", dev, "-n", clear_dev],
+                        { superuser: true });
 }
 
-export async function unlock_with_type(client, block, passphrase, passphrase_type) {
+export async function unlock_with_type(client, block, passphrase, passphrase_type, override_readonly) {
```
[the remark this answers is not in the document]
Comment (mvollmer) [reply pkg/storaged/crypto/keyslots.jsx:98]: `override_readonly` has three meanings:

 -  null / undefined: don't override, take readonly-nesss from /etc/crypttab
 -  true: make it readonly, regradless of what cryttab says
 - false: make it readwrite, regardless of what crypttab says

The default if not specified is "don't override".

**Review by jelly — APPROVED** [review]

**Pull request merged**